### ISSUE TITLE
Fail to Load a story Saved state

### ISSUE THREAD

**Issue by Tayx94** [inkle/ink#118]
Hi! I'm currently trying to save the story state and load it afterwards, but I'm facing a bit of a problem. I need to change scenes and be able to load the Story state to a "new Story()" made with the same ".json" file as the one used to save the state before.
To save the state I'm using PlayerPrefs, and while debuggin it it shows that the state is being saved correctly in json format. The problem arises when I try to load the state to the new Story, when Ink throws this error:

![image](https://cloud.githubusercontent.com/assets/16625646/16401088/b04e560e-3ce1-11e6-9950-048df7f19875.png)

I tried to look into the ink-engine-runtime code, but couldn't come to any clear conclusion.

Any kind of help would be greatly apreciatted!

**Comment by JosephLoe** [comment]
I'm glad i'm not the only person having this issue.
`
IndexOutOfRangeException: Array index is out of range.
Ink.Runtime.Path.set_componentsString (System.String value)
Ink.Runtime.Path..ctor (System.String componentsString)
Ink.Runtime.CallStack+Thread..ctor (Newtonsoft.Json.Linq.JToken jsonToken, Ink.Runtime.Story storyContext)
Ink.Runtime.CallStack.SetJsonToken (Newtonsoft.Json.Linq.JToken token, Ink.Runtime.Story storyContext)
Ink.Runtime.StoryState.set_jsonToken (Newtonsoft.Json.Linq.JToken value)
Ink.Runtime.StoryState.LoadJson (System.String json)
`
Above is the error i'm receiving, an index error.
Below is the code snippet causing the index error

`
    void Start ()
    {
        _inkStory = new Story(inkAsset.text);
        saveState = _inkStory.state.ToJson();
        CM.openConversation(_inkStory, saveState);
    }

    public void openConversation(Story storyObj, string storySaveState)
    {
        enableUI();
        storyObj.state.LoadJson(storySaveState);
        updateStory(storyObj, storySaveState);
    }
`

**Comment by nullObjectPtr** [comment]
Someone from Inkle may have a more technical explanation, so take what I say with a grain of salt. I haven't actually stepped through the code to figure out what's going on. 

Each time that I've encountered this error, the cause was attempting to load a save that was created with an older version of the Ink story. Try deleting the save, and saving a new one. You should be able to load that state back in without any trouble.

**Comment by joethephish** [comment]
Hrm, I haven't had a chance to look into it fully, but it's possible that there's a bug if you try to save the state before any choices have been made, so before calling `Continue()` for the first time. That might explain why @JosephLoe's code snippet doesn't work?

This is just a guess though, until I've had a chance to investigate properly!

**Comment by Tayx94** [comment]
I think that's the problem. I wanted to save the original story state to load it later, but without calling Continue() it always failed. Do you know if it's going to be fixed soon? Thanks!

**Comment by joethephish** [comment]
Sorry, I haven't had a chance to look at this yet, and will be quite busy for the next few months. But will look at it at some point, if no-one else does!

**Comment by nullObjectPtr** [comment]
I did some digging. I loaded up one of my ink files and compared two saves files; One, before Continue() was called and one after a single call to Continue(). Figured they couldn't be too different and the problem would reveal pretty quickly.

Found that the callstack thread differed a bit and looked into why.  

// Working Version
```
	"callstackThreads": {
		"threads": [{
			"callstack": [{
				"exp": false,
				"type": 0,
				"temp": {}
			}],
			"threadIndex": 0,
			"previousContentObject": "feelin_it.0.2.5"
		}],
		"threadCounter": 0
	},
```
// Non working version
```
	"callstackThreads": {
		"threads": [{
			"callstack": [{
				"cPath": "",
				"idx": 0,
				"exp": false,
				"type": 0,
				"temp": {}
			}],
			"threadIndex": 0
		}],
		"threadCounter": 0
	},
```

Before continue is called, the first callstack thread has contentContainer set to the story's mainContentContainer, whoose path is the empty string "".

This is serialized out. But on deserialization, the parser is assuming that cPath is not an empty string. In Path.cs the setter for the property ComponentsString attempts to access the first character of the path string, without checking it's length.

```
    if (componentsStr[0] == '.')
    {
        isRelative = true;
        componentsStr = componentsStr.Substring(1);
    }
```

The right fix depends on whether or not the empty string is a valid value for a contentPath. From what I can gather, the root content container's path is "", ... so ... it is. But I could have that wrong. Could someone verify that?

Locally, in the setter, after the call to components.Clear().

I just wrapped the rest of the logic in a 
```
if(string.IsNullOrEmpty(value))
{
	....	
}
```
And things seem to be fine. 


Unrelated, but one thing that popped out at me as pontentially problematic was the update to the boolean field isRelative. Which is only ever set to true if the path is relative, but not false if it isn't. If a path object is not treated as read-only after instantiation, subsequent calls to the contentPath setter could incorrectly set the isRelative bool to true. 

i.e.
```
var p = new Path()
p.componentsString = ".someRelativePath" // isRelative is true
p.componentsString = "someNonRelativePath" // isRelative is still true
```
isRelative also has a public setter, but since it's derived from the componentPath, it could be made private no?

**Comment by joethephish** [comment]
Thanks @nullObjectPtr, that was helpful! I've now committed a fix: f95f37cabe28c9a147196beca344c41edaf51c78

Paths can indeed be the empty string, and that's something that I allowed/fixed relatively recently, but it seems that I hadn't hit this case.

The `componentsString` property on `Path` isn't really meant to be set more than once, but I've tidied up the logic there too, and made it private. `Path` is really designed to be used as an immutable class.
